scripts/feature_tests/upgrade/common.sh

```diff
@@ -0,0 +1,486 @@
+    for i in {1..1800};do
+        ug_started=$(kubectl get bmh -n metal3 | awk 'NR>1' | grep -c 'provisioning')
+        if [[ "${ug_started}" -gt "0" ]]; then
+            break
+        fi
+    done
+    while [ ${ug_ongoing} -eq 1 ]; do
+        running_machines=$(kubectl get machines -n metal3 | awk 'NR>1'| grep -c 'Running')
+        other_machines=$(kubectl get machines -n metal3 | awk 'NR>1'| grep -cv 'Running')
+            if [[ "${other_machines}" -eq "0" && "${running_machines}" -eq "${NUM_OF_MASTER_REPLICAS}" ]]; then
+                provisioned_bmhs=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -c 'provisioned')
+                other_bmhs=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -cv 'provisioned')
+                ready_bmhs=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -c 'ready')
+                if [[ "${other_bmhs}" -eq "${exp_count}" && "${provisioned_bmhs}" -eq "${NUM_OF_MASTER_REPLICAS}" && "${ready_bmhs}" -eq "${exp_count}" ]]; then
+                    echo ''
+                    echo "Successfully upgraded the k8s version of ${NUM_OF_MASTER_REPLICAS} control plane nodes"
+                    ug_ongoing=0
+                fi
+            fi
+            # Upgrade progress indicator
+            echo -n "-"
+            sleep 10
+            # shellcheck disable=SC2004
+            node_c=$(($node_c+1))
+            if [[ "${node_c}" -eq "7200" ]]; then
+                ug_ongoing=2
+                echo "Upgrade failed, resource(s) are hanging."
+                deprovision_cluster
+                wait_for_cluster_deprovisioned
+                break
+            fi
+    done
+
+    # Upgrade succeeded, test connectivity
+    if [[ "${ug_ongoing}" -eq "0" ]]; then
+        # pick the first node and test ssh on it
+        NEW_NODE_NAME=$(kubectl get bmh -n metal3 | awk 'NR>1' | grep 'prov' | grep 'control' | awk '{{print $1}}' | head -1)
+        if [ -n "${NEW_NODE_NAME}" ]; then
+            new_node_ip=$(sudo virsh net-dhcp-leases baremetal | grep "${NEW_NODE_NAME}" | awk '{{print $5}}' | cut -f1 -d'/')
+            for i in {1..1800};do
+                ssh -o "UserKnownHostsFile=/dev/null" -o PasswordAuthentication=no -o "StrictHostKeyChecking no" "${UPGRADE_USER}@${new_node_ip}" -- kubectl version 2>&1 /dev/null
+                # shellcheck disable=SC2181
+                if [[ "$?" == "0" ]]; then
+                    echo ''
+                    echo "Successfully connected over ssh to an upgraded control plane node"
+                    server_version=$(ssh -o "UserKnownHostsFile=/dev/null" -o "StrictHostKeyChecking no" "${UPGRADE_USER}@${new_node_ip}" -- kubectl version --short | grep Server)
+                    client_version=$(ssh -o "UserKnownHostsFile=/dev/null" -o "StrictHostKeyChecking no" "${UPGRADE_USER}@${new_node_ip}" -- kubectl version --short | grep Client)
+                    echo "${server_version}"
+                    echo "${client_version}"
```

Comment:
this whole connection part is also repeated throughout the file and could be gathered in a function if properly parameterized